services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/endpoints/S3EndpointResolverContext.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.endpoints;
+
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Configuration;
+
+/**
+ * Contains the information needed to resolve S3 endpoints.
+ */
+@SdkInternalApi
+public final class S3EndpointResolverContext {
```

Comment:
Since we're following the classic SDK pojo pattern here, we should also add a 'toBuilder()' method.

Reply:
Yes. Good that you commented on this class, as I found two unused fields in the builder.

Reply:
toBuilder needs a test too. Usually it's a fairly standard one where we just build() then toBuilder.build() and compare the 2nd object with the first.